File: VisualBasic/PetShop/Source/PetShop.Tests/PetShop.Tests.OF.ParameterizedSQL/TestUtility.vb
Imports System
Imports System.Text

Public Class TestUtility
    Private Shared _instance As TestUtility

    'Random class used when generating strings.  Using these classes will
    'guarantee that multiple calls to these functions should return 
    'random values.
    Private ReadOnly _stringRandom As New Random()
    Private ReadOnly _dateRandom As New Random()
    Private ReadOnly _numberRandom As New Random()

    Private ReadOnly _minDate As New DateTime(1990, 1, 1)
    Private ReadOnly _maxDate As New DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day)

    ''' <summary>
    ''' Private constructor for Singleton
    ''' </summary>
    Private Sub New()

    End Sub

    ''' <summary>
    ''' Get the instance of the TestUtility class
    ''' </summary>
    Public Shared ReadOnly Property Instance() As TestUtility
        Get
            If _instance Is Nothing Then
                _instance = New TestUtility()
            End If

            Return _instance
        End Get
    End Property

    ''' <summary>
    ''' Generates a random number between 0 and int.MaxValue.
    ''' </summary>
    ''' <returns>Random Number</returns>
    Public Function RandomNumber() As Integer
        Return RandomNumber(0, Integer.MaxValue)
    End Function

    ''' <summary>
    ''' Generates a random number between the given bounds.
    ''' </summary>
    ''' <param name="min">lowest bound</param>
    ''' <param name="max">highest bound</param>
    ''' <returns>Random Number</returns>
    Public Function RandomNumber(ByVal min As Integer, ByVal max As Integer) As Integer
        Return _numberRandom.[Next](min, max)
    End Function

    ''' <summary>
    ''' Generates a random string with the given length
    ''' </summary>
    ''' <param name="size">Size of the string</param>
    ''' <param name="lowerCase">If true, generate lowercase string</param>
    ''' <returns>Random string</returns>
    ''' <remarks>Mahesh Chand  - http://www.c-sharpcorner.com/Code/2004/Oct/RandomNumber.asp</remarks>
    Public Function RandomString(ByVal size As Integer, ByVal lowerCase As Boolean) As String
        Dim builder As New StringBuilder()

        For i As Integer = 0 To size - 1
            Dim ch As Char = Convert.ToChar(Convert.ToInt32(26 * _stringRandom.NextDouble() + 65))
            builder.Append(ch)
        Next

        Dim retVal = If(lowerCase, builder.ToString().ToLower(), builder.ToString())

        Return retVal
    End Function

    ''' <summary>
    ''' Returns a random date between the default min and max dates
    ''' </summary>
    ''' <returns></returns>
    Public Function RandomDate() As DateTime
        Return RandomDate(_minDate, _maxDate)
    End Function

    ''' <summary>
    ''' Returns a random date between the dates you pass in
    ''' </summary>
    ''' <param name="minDate">Min date to return</param>
    ''' <param name="maxDate">Max date to return</param>
    ''' <returns></returns>
    Public Function RandomDate(ByVal minDate As DateTime, ByVal maxDate As DateTime) As DateTime
        'Get the total days between the 2 dates
        Dim totalDays As Integer = CInt(maxDate.Subtract(minDate).TotalDays)

        'Pick a random date in between
        Dim randomDays As Integer = _dateRandom.[Next](0, totalDays)

        'Return the random day.
        Return minDate.AddDays(randomDays)
    End Function

    ''' <summary>
    ''' Returns a random date and time between the default datess
    ''' </summary>
    ''' <returns></returns>
    Public Function RandomDateTime() As DateTime
        Return RandomDateTime(_minDate, _maxDate)
    End Function

    ''' <summary>
    ''' Get a random DateTime with a random Time
    ''' </summary>
    ''' <param name="minDate">Min datetime to return</param>
    ''' <param name="maxDate">Max datetime to return</param>
    ''' <returns></returns>
    Public Function RandomDateTime(ByVal minDate As DateTime, ByVal maxDate As DateTime) As DateTime
        'Get the total seconds between the 2 dates
        'Careful of overflow here
        Dim totalSeconds As Integer = CInt(maxDate.Subtract(minDate).TotalSeconds)

        'Pick a random date in between
        Dim randomSeconds As Integer = _dateRandom.[Next](0, totalSeconds)

        'Return the random date.
        Return minDate.AddSeconds(randomSeconds)
    End Function

    ''' <summary>
    ''' Returns a random boolean value
    ''' </summary>
    ''' <returns>Random Boolean</returns>
    Public Function RandomBoolean() As Boolean
        'if the second is odd, return True
        Return ((DateTime.Now.Second Mod 2) > 0)
    End Function

    ''' <summary>
    ''' Returns a random character
    ''' </summary>
    ''' <returns>Random Character</returns>
    Public Function RandomChar() As Char
        Return Convert.ToChar(Convert.ToInt32(26 * _stringRandom.NextDouble() + 65))
    End Function

    ''' <summary>
    ''' Return a random byte between 0 and byte.MaxValue;
    ''' </summary>
    ''' <returns></returns>
    Public Function RandomByte() As Byte
        Return RandomByte(0, Byte.MaxValue)
    End Function

    ''' <summary>
    ''' Return a random byte between the values specified
    ''' </summary>
    ''' <param name="min">Min value</param>
    ''' <param name="max">Max value</param>
    ''' <returns>Random Byte</returns>
    Public Function RandomByte(ByVal min As Byte, ByVal max As Byte) As Byte
        Return CByte(RandomNumber(min, max))
    End Function

    ''' <summary>
    ''' Return a random short between 0 and byte.MaxValue;
    ''' </summary>
    ''' <returns></returns>
    Public Function RandomShort() As Short
        Return RandomShort(0, Short.MaxValue)
    End Function

    ''' <summary>
    ''' Return a random short between the values specified
    ''' </summary>
    ''' <param name="min">Min value</param>
    ''' <param name="max">Max value</param>
    ''' <returns>Random short</returns>
    Public Function RandomShort(ByVal min As Short, ByVal max As Short) As Short
        Return CShort(RandomNumber(min, max))
    End Function
End Class

File: CSharp/PetShop.Silverlight/Source/PetShop.Tests/PetShop.Tests.ObjF.ParameterizedSQL/TestUtility.cs
using System;
using System.Text;

namespace PetShop.Tests.ObjF.ParameterizedSQL
{
    public class TestUtility
    {
        private static TestUtility _instance;

        //Random class used when generating strings.  Using these classes will
        //guarantee that multiple calls to these functions should return 
        //random values.
        private readonly Random _stringRandom = new Random();
        private readonly Random _dateRandom = new Random();
        private readonly Random _numberRandom = new Random();

        private readonly DateTime _minDate = new DateTime(1990, 1, 1);
        private readonly DateTime _maxDate = new DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day);

        /// <summary>
        /// Private constructor for Singleton
        /// </summary>
        private TestUtility()
        {

        }

        /// <summary>
        /// Get the instance of the TestUtility class
        /// </summary>
        public static TestUtility Instance
        {
            get
            {
                if (_instance == null)
                    _instance = new TestUtility();

                return _instance;
            }
        }

        /// <summary>
        /// Generates a random number between 0 and int.MaxValue.
        /// </summary>
        /// <returns>Random Number</returns>
        public int RandomNumber()
        {
            return RandomNumber(0, int.MaxValue);
        }

        /// <summary>
        /// Generates a random number between the given bounds.
        /// </summary>
        /// <param name="min">lowest bound</param>
        /// <param name="max">highest bound</param>
        /// <returns>Random Number</returns>
        public int RandomNumber(int min, int max)
        {
            return _numberRandom.Next(min, max);
        }

        /// <summary>
        /// Generates a random string with the given length
        /// </summary>
        /// <param name="size">Size of the string</param>
        /// <param name="lowerCase">If true, generate lowercase string</param>
        /// <returns>Random string</returns>
        /// <remarks>Mahesh Chand  - http://www.c-sharpcorner.com/Code/2004/Oct/RandomNumber.asp</remarks>
        public string RandomString(int size, bool lowerCase)
        {
            StringBuilder builder = new StringBuilder();

            for (int i = 0; i < size; i++)
            {
                char ch = Convert.ToChar(Convert.ToInt32(26 * _stringRandom.NextDouble() + 65));
                builder.Append(ch);
            }

            var retVal = lowerCase ? builder.ToString().ToLower() : builder.ToString();

            return retVal;
        }

        /// <summary>
        /// Returns a random date between the default min and max dates
        /// </summary>
        /// <returns></returns>
        public DateTime RandomDate()
        {
            return RandomDate(_minDate, _maxDate);
        }

        /// <summary>
        /// Returns a random date between the dates you pass in
        /// </summary>
        /// <param name="minDate">Min date to return</param>
        /// <param name="maxDate">Max date to return</param>
        /// <returns></returns>
        public DateTime RandomDate(DateTime minDate, DateTime maxDate)
        {
            //Get the total days between the 2 dates
            int totalDays = (int)maxDate.Subtract(minDate).TotalDays;

            //Pick a random date in between
            int randomDays = _dateRandom.Next(0, totalDays);

            //Return the random day.
            return minDate.AddDays(randomDays);
        }

        /// <summary>
        /// Returns a random date and time between the default datess
        /// </summary>
        /// <returns></returns>
        public DateTime RandomDateTime()
        {
            return RandomDateTime(_minDate, _maxDate);
        }

        /// <summary>
        /// Get a random DateTime with a random Time
        /// </summary>
        /// <param name="minDate">Min datetime to return</param>
        /// <param name="maxDate">Max datetime to return</param>
        /// <returns></returns>
        public DateTime RandomDateTime(DateTime minDate, DateTime maxDate)
        {
            //Get the total seconds between the 2 dates
            //Careful of overflow here
            int totalSeconds = (int)maxDate.Subtract(minDate).TotalSeconds;

            //Pick a random date in between
            int randomSeconds = _dateRandom.Next(0, totalSeconds);

            //Return the random date.
            return minDate.AddSeconds(randomSeconds);
        }

        /// <summary>
        /// Returns a random boolean value
        /// </summary>
        /// <returns>Random Boolean</returns>
        public bool RandomBoolean()
        {
            //if the second is odd, return True
            return ((DateTime.Now.Second % 2) > 0);
        }

        /// <summary>
        /// Returns a random character
        /// </summary>
        /// <returns>Random Character</returns>
        public char RandomChar()
        {
            return Convert.ToChar(Convert.ToInt32(26 * _stringRandom.NextDouble() + 65));
        }

        /// <summary>
        /// Return a random byte between 0 and byte.MaxValue;
        /// </summary>
        /// <returns></returns>
        public byte RandomByte()
        {
            return RandomByte(0, byte.MaxValue);
        }

        /// <summary>
        /// Return a random byte between the values specified
        /// </summary>
        /// <param name="min">Min value</param>
        /// <param name="max">Max value</param>
        /// <returns>Random Byte</returns>
        public byte RandomByte(byte min, byte max)
        {
            return (byte)RandomNumber(min, max);
        }

        /// <summary>
        /// Return a random short between 0 and byte.MaxValue;
        /// </summary>
        /// <returns></returns>
        public short RandomShort()
        {
            return RandomShort(0, short.MaxValue);
        }

        /// <summary>
        /// Return a random short between the values specified
        /// </summary>
        /// <param name="min">Min value</param>
        /// <param name="max">Max value</param>
        /// <returns>Random short</returns>
        public short RandomShort(short min, short max)
        {
            return (short)RandomNumber(min, max);
        }
    }
}
